### PR TITLE
Fix show & close for fast code close of dialog

### PR DESCRIPTION
**Pull request by rockResolve**
For progress dialogs: allow fast code close. Fix for #137

Ensure show & close always complete
add options.showDialogTimeout
return model.showDialogPromise

Includes documentation changes and sample demos.
Sample demo includes faulty usage button, to show the problem if showDialogPromise is not used.

### CODE REVIEW

**Comment by EisenbergEffect** [comment]
@jedd-ahyoung Can you review please?

File: sample/src/progress.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<template>
+  <require from='./progress.css'></require>
+    <ai-dialog>
+        <ai-dialog-body>
+            <div class='blockPageDialog'>
```
Comment (PWKad) [inline sample/src/progress.html]: Can you fix the tabs so that they are all two spaces?  (README.md also has mixed spacing)

**Comment by jedd-ahyoung** [comment]
@rockResolve Thanks for the PR! I'll review it during the week. In the meantime, can you amend the commit to follow the [Aurelia commit guidelines](https://github.com/DurandalProject/about/blob/master/CONTRIBUTING.md#commit)?

**Comment by rockResolve** [comment]
@PWKad @jedd-ahyoung 
I've amended the commit to fix your catches as requested.

**Comment by jedd-ahyoung** [comment]
@EisenbergEffect  I've taken a look at this and it looks pretty good. @rockResolve I do have a question, though - explained here (https://github.com/aurelia/dialog/issues/137), you mentioned that closing the dialog before the open is finished causes issues. The solution in the PR fixes this, but requires client code to change (by adding the transition promise). I'm wondering if we can remove that necessity by simply storing the promise that the `showDialog` method returns - we can have the `hideDialog` method in the renderer only fire after that promise has resolved. (Something like - `return showDialogPromise.then(() => { return new Promise(...) });`) This would avoid having to change the interface to expose the transition. Do you think this would work for your use case while fixing the general issue?

Let me know what you think.

**Comment by rockResolve** [comment]
@jedd-ahyoung That would improve things generally. 

For me I have multiple sequential progress dialogs (i.e. get generic startup data, then first view's data). So when opening a possibly open progress dialog I need to know the dialog's state, e.g is it opening. I am using the showDialogPromise for this. 

With your suggestion I could just go ahead and close anytime, using the dailogController.close returned promise to tell if the dialog is closed yet (or to wait for it to close before opening again). So your suggestion would work for me.

I cant think of a scenario where anyone would want to know for sure the dialog is fully open except to ensure it closes. Changing messages while a dialog would be handled by binding. Even timed dialogs (e.g. close after 2 minutes) wouldnt care if the timer was before or after the 200 ms transition.

Sounds like a good option as it removes the promise from the interface.

**Comment by EisenbergEffect** [comment]
In light of the conversation here, I'm going to close this PR. Can someone submit a new PR with the proposed change. @jedd-ahyoung is that something you have time to do?

**Comment by jedd-ahyoung** [comment]
@EisenbergEffect Sure. @rockResolve has actually done all of the legwork here. It'd just be a matter of removing the promise from `_settings.model` and storing it in the renderer.

**Comment by EisenbergEffect** [comment]
Ok, sounds good. When you get a chance, just submit the new PR and we can merge. Are the other dialog PRs good to go? Should I get them merged first?

**Comment by jedd-ahyoung** [comment]
No. The refactor has a small fix that I need to do, and I need to try to
fix the test coverage. The renderer changes for scrolling, etc, are still
suffering from the click issue that we'd discussed before.

2016-05-04 11:50 GMT-04:00 Rob Eisenberg <notifications@github.com>:

**Comment by flieks** [comment]
Hi,

How can i close the dialog from within the viewmodel that launched the dialog (and received the promise)?

Greetings

**Comment by rockResolve** [comment]
controller.close(), controller is made available during component activation.

For a summary have a look at the PR ReadMe.md, heading "Force close from code - Progress dialogs". 

For the detail see PR's sample folder. Look at app.js fastestProgress method. It uses closeDialog created by progress.js

**Comment by flieks** [comment]
Sorry @rockResolve but what is PR ?
I only find this [readme.md](https://github.com/aurelia/dialog/blob/master/README.md)

**Comment by rockResolve** [comment]
PR = Pull Request
See https://help.github.com/articles/using-pull-requests/

**Comment by flieks** [comment]
Right i see, sorry.
But i don't need a progressbar, just want to close the dialog from the parent VM.

**Comment by valichek** [comment]
I think this issue is closely related to closing dialog internally and having no access to dialog controller when opened. Currently it is not possible to get promise for dialog opening (transitions, animations, etc) outside. And I still think that my suggestion for API changes (https://github.com/aurelia/dialog/issues/96#issuecomment-219399087) is the only way to implement such scenarios but might be I'm wrong, so would be happy to get some working example.

**Comment by valichek** [comment]
But the suggested behavior could be optional depending on dialogOptions passed

**Pull request closed without merging**